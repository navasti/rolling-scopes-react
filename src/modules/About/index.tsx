import { Layout } from 'modules';
import * as S from './styled';
import React from 'react';

type Props = {
  componentName: string;
  location: string;
};

export const About = ({ componentName, location }: Props) => {
  return (
    <Layout location={location} componentName={componentName}>
      <S.CommonView>
        <S.CommonMessage>About us page</S.CommonMessage>
      </S.CommonView>
    </Layout>
  );
};
