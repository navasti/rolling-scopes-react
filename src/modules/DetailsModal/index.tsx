import { PokemonDetails, PokemonMoveDetails, PokemonTypeDetails } from 'types';
import { Modal } from 'components';
import { forwardRef } from 'react';
import * as S from './styled';
import {
  PokemonModalContent,
  MoveModalContent,
  TypeModalContent,
} from 'modules/DetailsModal/components';

type Props = {
  selectedType?: PokemonTypeDetails;
  selectedMove?: PokemonMoveDetails;
  selectedPokemon?: PokemonDetails;
  handleCloseModal: () => void;
  isModalOpened: boolean;
};

export const DetailsModal = forwardRef<HTMLDivElement, Props>((props, ref) => {
  const { handleCloseModal, isModalOpened, selectedMove, selectedPokemon, selectedType } = props;
  return (
    <>
      {selectedPokemon ? (
        <Modal handleCloseModal={handleCloseModal} opened={isModalOpened} ref={ref}>
          <S.ModalTitle>{selectedPokemon.name}</S.ModalTitle>
          <PokemonModalContent selectedPokemon={selectedPokemon} />
        </Modal>
      ) : selectedMove ? (
        <Modal handleCloseModal={handleCloseModal} opened={isModalOpened} ref={ref}>
          <S.ModalTitle>{selectedMove.name}</S.ModalTitle>
          <MoveModalContent selectedMove={selectedMove} />
        </Modal>
      ) : selectedType ? (
        <Modal handleCloseModal={handleCloseModal} opened={isModalOpened} ref={ref}>
          <S.ModalTitle>{selectedType.name}</S.ModalTitle>
          <TypeModalContent selectedType={selectedType} />
        </Modal>
      ) : null}
    </>
  );
});
