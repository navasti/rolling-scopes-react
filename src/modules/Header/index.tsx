import { LocationBadge } from 'modules/Header/components';
import { Link } from 'react-router-dom';
import * as S from './styled';

type Props = {
  componentName: string;
  location: string;
};

export const Header = ({ componentName, location }: Props) => {
  return (
    <S.Header>
      <S.Nav role="navigation">
        <S.LinkWrapper>
          <Link role="link" to="/">
            Home
          </Link>
        </S.LinkWrapper>
        <S.LinkWrapper>
          <Link role="link" to="/about">
            About
          </Link>
        </S.LinkWrapper>
        <S.LinkWrapper>
          <Link role="link" to="/form">
            Form
          </Link>
        </S.LinkWrapper>
        <S.LinkWrapper>
          <Link role="link" to="/invalid">
            Invalid
          </Link>
        </S.LinkWrapper>
      </S.Nav>
      <LocationBadge componentName={componentName} location={location} />
    </S.Header>
  );
};
