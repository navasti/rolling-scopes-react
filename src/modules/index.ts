export { LocationBadge } from './Header/components';
export { DetailsModal } from './DetailsModal';
export { SearchPage } from './SearchPage';
export { NotFound } from './NotFound';
export { Header } from './Header';
export { Layout } from './Layout';
export { About } from './About';
export { Form } from './Form';
export { App } from './App';
