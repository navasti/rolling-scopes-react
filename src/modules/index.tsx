export { App } from './App';
export { Form } from './Form';
export { About } from './About';
export { Layout } from './Layout';
export { Header } from './Header';
export { NotFound } from './NotFound';
export { SearchPage } from './SearchPage';
export { LocationBadge } from './Header/components';
export { Card, SearchBar } from './SearchPage/components';
