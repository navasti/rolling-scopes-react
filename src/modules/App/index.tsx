import { SearchPage, About, NotFound, Form } from 'modules';
import { Navigate, Route, Routes } from 'react-router-dom';
import React from 'react';

export const App = () => {
  return (
    <Routes>
      <Route path="/" element={<SearchPage componentName="SearchPage" location="/" />} />
      <Route path="/about" element={<About componentName="About" location="/about" />} />
      <Route path="/form" element={<Form componentName="Form" location="/form" />} />
      <Route path="/404" element={<NotFound componentName="NotFound" location="/404" />} />
      <Route path="*" element={<Navigate to="/404" replace />} />
    </Routes>
  );
};
