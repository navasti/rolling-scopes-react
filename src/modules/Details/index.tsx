import { MoveCardDetails, PokemonCardDetails, TypeCardDetails } from './components';
import { Link, useParams } from 'react-router-dom';
import { useEffect, useState } from 'react';
import { useGlobalContext } from 'contexts';
import { Layout } from 'modules';
import * as S from './styled';
import {
  AvailableCardDetails,
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
} from 'types';

type Props = {
  componentName: string;
  location: string;
};

type MatchedPokemon = {
  resourceType: AvailableCardDetails.pokemon;
  data: PokemonDetails;
};

type MatchedMove = {
  resourceType: AvailableCardDetails.move;
  data: PokemonMoveDetails;
};

type MatchedType = {
  resourceType: AvailableCardDetails.type;
  data: PokemonTypeDetails;
};

type MatchedData = MatchedPokemon | MatchedMove | MatchedType;

export const Details = ({ componentName, location }: Props) => {
  const [matchedData, setMatchedData] = useState<MatchedData>();

  const { id, resourceType } = useParams();

  const {
    state: { allDataResults },
  } = useGlobalContext();

  useEffect(() => {
    const { move, pokemon, type } = AvailableCardDetails;
    if (id && resourceType === pokemon) {
      const found = allDataResults.pokemons.find((pokemon) => pokemon.id === Number(id));
      found && setMatchedData({ resourceType, data: found });
    }
    if (id && resourceType === move) {
      const found = allDataResults.moves.find((move) => move.id === Number(id));
      found && setMatchedData({ resourceType, data: found });
    }
    if (id && resourceType === type) {
      const found = allDataResults.types.find((type) => type.id === Number(id));
      found && setMatchedData({ resourceType, data: found });
    }
  }, [resourceType, id, allDataResults]);

  return (
    <Layout location={location} componentName={componentName}>
      <S.DetailsView>
        {matchedData ? (
          <>
            {matchedData.resourceType === AvailableCardDetails.pokemon && (
              <PokemonCardDetails pokemon={matchedData.data} />
            )}
            {matchedData.resourceType === AvailableCardDetails.move && (
              <MoveCardDetails move={matchedData.data} />
            )}
            {matchedData.resourceType === AvailableCardDetails.type && (
              <TypeCardDetails type={matchedData.data} />
            )}
          </>
        ) : (
          <S.DetailsInformation>
            {id && resourceType ? (
              <>
                Could not found detailed data for {resourceType} with id {id}
              </>
            ) : (
              <>
                No enugh information provided. Get back to the search page in order to fetch data
                and select specific pokemon to see its details.
              </>
            )}
          </S.DetailsInformation>
        )}
        <S.ReturnButton>
          <Link to="/">Go to search page</Link>
        </S.ReturnButton>
      </S.DetailsView>
    </Layout>
  );
};
