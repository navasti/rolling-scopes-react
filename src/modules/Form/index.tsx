import React, { createRef, FormEvent, RefObject } from 'react';
import { CheckFields, CustomPokemon, ErrorsObject, TextFields } from 'types';
import { Layout } from 'modules';
import * as S from './styled';
import { uuid } from 'utils';
import {
  BirthdayField,
  ErrorMessage,
  ConsentField,
  GenderField,
  AvatarField,
  ShinyField,
  TypeField,
  NameField,
  FormCard,
} from 'modules/Form/components';
import { FIELDS_VALIDATION_BY_NAME, FEMALE, MALE, ErrorMessages } from 'appConstants';

type Props = {
  componentName: string;
  location: string;
};

type State = {
  pokemons: Array<CustomPokemon & { id: string }>;
  initialEnter: boolean;
  errors: ErrorsObject;
};

export class Form extends React.Component<Props, State> {
  femaleInputRef: RefObject<HTMLInputElement>;
  successMessage: RefObject<HTMLSpanElement>;
  birthdayField: RefObject<HTMLInputElement>;
  isShinyField: RefObject<HTMLInputElement>;
  consentField: RefObject<HTMLInputElement>;
  maleInputRef: RefObject<HTMLInputElement>;
  avatarField: RefObject<HTMLInputElement>;
  typeField: RefObject<HTMLSelectElement>;
  nameField: RefObject<HTMLInputElement>;

  constructor(props: Props) {
    super(props);
    this.state = {
      pokemons: [],
      initialEnter: true,
      errors: {
        birthday: null,
        consent: null,
        gender: null,
        name: null,
        type: null,
      },
    };
    this.successMessage = createRef();
    this.femaleInputRef = createRef();
    this.birthdayField = createRef();
    this.consentField = createRef();
    this.isShinyField = createRef();
    this.maleInputRef = createRef();
    this.avatarField = createRef();
    this.typeField = createRef();
    this.nameField = createRef();
  }

  removeError = (field: TextFields | CheckFields) => {
    if (this.state.initialEnter) this.setState({ initialEnter: false });
    const errors = { ...this.state.errors };
    errors[field] = null;
    this.setState({ errors });
  };

  validate = (fields: Array<React.RefObject<HTMLInputElement | HTMLSelectElement>>) => {
    const errors = { ...this.state.errors };
    const textFields = FIELDS_VALIDATION_BY_NAME.TEXT.map((textField) =>
      fields.find((field) => field.current?.name === textField)
    );
    const checkFields = FIELDS_VALIDATION_BY_NAME.CHECK.map((checkField) =>
      fields.filter((field) => field.current?.name === checkField)
    );
    checkFields.forEach((checkField) => {
      const valid = checkField.some((field) => (field.current as HTMLInputElement)?.checked);
      const name = checkField[0]?.current?.name
        ? (checkField[0].current.name as CheckFields)
        : null;
      if (!valid && name) errors[name] = ErrorMessages[name];
    });
    textFields.forEach((textField) => {
      const name = textField?.current?.name ? (textField.current.name as TextFields) : null;
      const value = textField?.current?.value;
      const valid = value && value.trim().length >= 2;
      if (!valid && name) errors[name] = ErrorMessages[name];
    });
    return errors;
  };

  prepareCustomPokemon = () => {
    const shiny = !!this.isShinyField.current?.checked;
    const female = this.femaleInputRef?.current?.checked;
    const file = this.avatarField.current?.files?.[0] || null;
    const customPokemon: CustomPokemon & { id: string } = {
      birthday: this.birthdayField?.current?.value || '',
      type: this.typeField?.current?.value || '',
      name: this.nameField?.current?.value || '',
      gender: female ? FEMALE : MALE,
      avatar: file,
      id: uuid(),
      shiny,
    };
    return customPokemon;
  };

  handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const fields = [
      this.nameField,
      this.typeField,
      this.avatarField,
      this.isShinyField,
      this.consentField,
      this.birthdayField,
      this.maleInputRef,
      this.femaleInputRef,
    ];
    const errors = this.validate(fields);
    if (Object.values(errors).some((error) => error != null)) {
      this.setState({ errors });
    } else {
      const customPokemon = this.prepareCustomPokemon();
      this.setState({
        pokemons: [...this.state.pokemons, customPokemon],
      });
      fields.forEach((field) => {
        const current = field.current;
        if (current) {
          if (current.type === 'radio' || current.type === 'checkbox') {
            (current as HTMLInputElement).checked = false;
          } else current.value = '';
        }
      });
    }
  };

  render() {
    const { location, componentName } = this.props;
    return (
      <Layout location={location} componentName={componentName}>
        <S.CommonView>
          <S.FormHeading>Create custom pokemon!</S.FormHeading>
          <S.Form onSubmit={this.handleSubmit}>
            <NameField
              onChange={this.removeError}
              errors={this.state.errors}
              ref={this.nameField}
            />
            <TypeField
              onChange={this.removeError}
              errors={this.state.errors}
              ref={this.typeField}
            />
            <S.GenderWrapper>
              <S.RadioWrapper>
                *Gender
                <S.RadioFields>
                  <GenderField
                    name={FEMALE}
                    ref={this.femaleInputRef}
                    onChange={this.removeError}
                  />
                  <GenderField name={MALE} ref={this.maleInputRef} onChange={this.removeError} />
                </S.RadioFields>
              </S.RadioWrapper>
              <ErrorMessage
                visible={!!this.state.errors.gender}
                message={this.state.errors.gender}
              />
            </S.GenderWrapper>
            <BirthdayField
              ref={this.birthdayField}
              errors={this.state.errors}
              onChange={this.removeError}
            />
            <AvatarField ref={this.avatarField} />
            <ShinyField ref={this.isShinyField} />
            <ConsentField
              ref={this.consentField}
              errors={this.state.errors}
              onChange={this.removeError}
            />
            <S.SubmitButton
              disabled={
                this.state.initialEnter ||
                Object.values(this.state.errors).some((error) => error != null)
              }
              type="submit"
            >
              Submit
            </S.SubmitButton>
            <S.SuccessMessage ref={this.successMessage}>Pokemon added successfuly</S.SuccessMessage>
          </S.Form>
          <hr />
          <S.CardsGrid>
            {this.state.pokemons.map((pokemon) => (
              <FormCard key={pokemon.id} customPokemon={pokemon} />
            ))}
          </S.CardsGrid>
        </S.CommonView>
      </Layout>
    );
  }
}
