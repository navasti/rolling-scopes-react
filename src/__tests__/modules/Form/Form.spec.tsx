import { Form } from 'modules';

describe('Form', () => {
  it('should', () => {});
});
