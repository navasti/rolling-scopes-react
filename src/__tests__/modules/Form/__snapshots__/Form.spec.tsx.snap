// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should be rendered and match snapshot 1`] = `
<div>
  <div>
    <p>
      <span
        data-testid="component-mock"
      >
        Test
      </span>
      <span
        data-testid="location-mock"
      >
        /test
      </span>
    </p>
    <main>
      <div
        class="sc-bczRLJ ckBqkA"
      >
        <span
          class="sc-gsnTZi KSJOv"
        >
          Create custom pokemon!
        </span>
        <form
          class="sc-dkzDqf jQzErX"
        >
          <label
            for="name"
          >
            <input
              data-testid="name-mock"
              id="name"
              type="text"
            />
          </label>
          <label
            for="type"
          >
            <select
              data-testid="type-mock"
              id="type"
            />
          </label>
          <div
            class="sc-gKXOVf bEMisw"
          >
            <div
              class="sc-iBkjds kGGQgE"
            >
              *Gender
              <div
                class="sc-ftvSup lbyCAw"
              >
                <label
                  for="gender"
                >
                  <input
                    data-testid="gender-mock"
                    id="gender"
                    type="radio"
                    value="female"
                  />
                </label>
                <label
                  for="gender"
                >
                  <input
                    data-testid="gender-mock"
                    id="gender"
                    type="radio"
                    value="male"
                  />
                </label>
              </div>
            </div>
            <span
              data-testid="message-mock"
              style="opacity: 0; text-align: left; color: red;"
            />
          </div>
          <label
            for="birth"
          >
            <input
              data-testid="birth-mock"
              id="birth"
              type="date"
            />
          </label>
          <label
            for="avatar"
          >
            <input
              data-testid="avatar-mock"
              id="avatar"
              type="file"
            />
          </label>
          <label
            for="shiny"
          >
            <input
              data-testid="shiny-mock"
              id="shiny"
              type="checkbox"
            />
          </label>
          <label
            for="consent"
          >
            <input
              data-testid="consent-mock"
              id="consent"
              type="checkbox"
            />
          </label>
          <button
            class="sc-eCYdqJ blCYfq"
            disabled=""
            type="submit"
          >
            Submit
          </button>
          <span
            data-testid="message-mock"
            style="opacity: 0; text-align: center; color: green;"
          >
            Pokemon added successfuly
          </span>
        </form>
        <hr />
        <div
          class="sc-jSMfEi iqEWfA"
        />
      </div>
    </main>
  </div>
</div>
`;
